### PR TITLE
Added VCP functionality trough USB (both RX & TX) trough DMA

### PR DESCRIPTION
**Pull request by allexoll**
This PR brings support for serial/vcp trough USB.

This works as is, but any remarks on structural impl are welcome!

Best,

Alexis

### CODE REVIEW

File: firmware/src/usb/mod.rs [hunk 1]
```diff
@@ -171,9 +171,13 @@ impl USB {
                 return r;
             }
 
-            // Discard data from the serial interface
             let mut buf = [0; DAP2_PACKET_SIZE as usize];
```
Comment (Disasm) [inline firmware/src/usb/mod.rs:174]:
```suggestion
            let mut buf = [0; VCP_PACKET_SIZE as usize];
```

File: firmware/src/vcp.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        }
+    }
+}
+
+pub struct VCP<'a> {
+    uart: usart::Instance,
+    pins: &'a Pins<'a>,
+    dma: &'a DMA,
+    rx_buffer: [u8; 256],
+    tx_buffer: [u8; 256],
```
Comment (Disasm) [inline firmware/src/vcp.rs:38]: USB packet size is 512 bytes, but this buffer is just 256 bytes: the firmware will panic after receiving a large packet.

**Comment by Disasm** [comment]
Thank you for the contribution! I added few notes, but other than that looks good to me.

**Review by Disasm — DISMISSED** [review]

**Review by adamgreig — COMMENTED** [review]
I haven't had a chance to test it but changes LGTM too.

**Comment by allexoll** [comment]
if you have a probe-rs on hand, you can just short tx&rx on the connector (last two pins on the opposite side of the led side) and test it using any serial tool like ` pio device monitor -b 115200    `

**Comment by Disasm** [comment]
I bet it will not withstand my USB serial loopback test due to backpressure issues, but still much better than nothing!

**Comment by Disasm** [comment]
Hmm, after the test if I run picocom multiple times I get some bytes every time even though I don't send any new data.

**Comment by allexoll** [comment]
huh yes and i sometime have a byte of delay even at boot but not with `--features=turbo`... I'm looking into it

**Comment by allexoll** [comment]
I fixed the data drop issue but i'm not very happy with the solution so far. The issue is that when we would receive a new VCP packet from the host with new data to pipe to the UART, there was no check to verify that we could transfer it. The solution I found is to not read the incoming data until we are able to pipe it. that adds a parameter to the `interrupt` function, which in itself is kind of weird, but that works as is, ive been able to test it with @Disasm's serial-threadmark:

```
read: 9999990 / 10000000
read: 9999996 / 10000000
read: 9999999 / 10000000
read completed
Time elapsed: 112.212907209s, throughput is 0.713 Mbit/s
```

with 921600 baud.

remarks welcome if you have an idea on how to make this nicer

**Comment by allexoll** [comment]
Sorry for the messy PR, I cleaned up the memory barrier for the UART TX DMA start, and removed the useless ones for UART RX

**Comment by korken89** [comment]
Hi,

Maybe replace the buffers with BBQueue?

**Comment by allexoll** [comment]
That could be un update for all DMA buffer

**Comment by allexoll** [comment]
ping @adamgreig

**Review by adamgreig — APPROVED** [review]
LGTM and I've tested loopback successfully. @Disasm, any thoughts? I'll merge this tomorrow otherwise.

**Comment by Disasm** [comment]
This PR looks good to me, but I don't have time to test it, will post my test results later. Let's merge it.

**Comment by adamgreig** [comment]
:+1: 

bors merge

**Pull request merged**